### PR TITLE
Improve compatibility of PCLConfig.cmake

### LINKED ISSUE #4661: [compile error] "can't found PCL because of wrong PCL_ROOT"
I installed PCL by `sudo apt install libpcl-dev`. and used it like this:
```cmake
cmake_minimum_required(VERSION 3.5)
project(test)

# Default to C99
if(NOT CMAKE_C_STANDARD)
  set(CMAKE_C_STANDARD 99)
endif()

# Default to C++14
if(NOT CMAKE_CXX_STANDARD)
  set(CMAKE_CXX_STANDARD 14)
endif()

if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
  add_compile_options(-Wall -Wextra -Wpedantic -Wno-unused-parameter)
endif()

# find dependencies
find_package(PCL REQUIRED)
```
**some errors occurred when I built the project:**
```cmake
CMake Error at /lib/x86_64-linux-gnu/cmake/pcl/PCLConfig.cmake:62 (message):
  PCL can not be found on this machine
Call Stack (most recent call first):
  /lib/x86_64-linux-gnu/cmake/pcl/PCLConfig.cmake:452 (pcl_report_not_found)
  CMakeLists.txt:37 (find_package)


-- Configuring incomplete, errors occurred!
```
**I'm trying to debug the cmake source code in `/lib/x86_64-linux-gnu/cmake/pcl/PCLConfig.cmake:418-454`, The cmake source is shown below:**
```cmake
find_package(PkgConfig QUIET)

file(TO_CMAKE_PATH "${PCL_DIR}" PCL_DIR)
if(WIN32 AND NOT MINGW)
# PCLConfig.cmake is installed to PCL_ROOT/cmake
  get_filename_component(PCL_ROOT "${PCL_DIR}" PATH)
else()
# PCLConfig.cmake is installed to PCL_ROOT/share/pcl-x.y
  get_filename_component(PCL_ROOT "${CMAKE_CURRENT_LIST_DIR}/../../../.." ABSOLUTE)
endif()

# check whether PCLConfig.cmake is found into a PCL installation or in a build tree
if(EXISTS "${PCL_ROOT}/include/pcl-${PCL_VERSION_MAJOR}.${PCL_VERSION_MINOR}/pcl/pcl_config.h")
  # Found a PCL installation
  # pcl_message("Found a PCL installation")
  set(PCL_CONF_INCLUDE_DIR "${PCL_ROOT}/include/pcl-${PCL_VERSION_MAJOR}.${PCL_VERSION_MINOR}")
  set(PCL_LIBRARY_DIRS "${PCL_ROOT}/lib/x86_64-linux-gnu")
  if(EXISTS "${PCL_ROOT}/3rdParty")
    set(PCL_ALL_IN_ONE_INSTALLER ON)
  endif()
elseif(EXISTS "${PCL_ROOT}/include/pcl/pcl_config.h")
  # Found a non-standard (likely ANDROID) PCL installation
  # pcl_message("Found a PCL installation")
  set(PCL_CONF_INCLUDE_DIR "${PCL_ROOT}/include")
  set(PCL_LIBRARY_DIRS "${PCL_ROOT}/lib")
  if(EXISTS "${PCL_ROOT}/3rdParty")
    set(PCL_ALL_IN_ONE_INSTALLER ON)
  endif()
elseif(EXISTS "${PCL_DIR}/include/pcl/pcl_config.h")
  # Found PCLConfig.cmake in a build tree of PCL
  # pcl_message("PCL found into a build tree.")
  set(PCL_CONF_INCLUDE_DIR "${PCL_DIR}/include") # for pcl_config.h
  set(PCL_LIBRARY_DIRS "${PCL_DIR}/lib/x86_64-linux-gnu")
  set(PCL_SOURCES_TREE "/build/pcl-gWGA5r/pcl-1.10.0+dfsg")
else()
  pcl_report_not_found("PCL can not be found on this machine")
endif()
```
**Through debugging， I found that the environment variable `PCL_ROOT` is set to `/` and `PCL_DIR` is set to `/lib/x86_64-linux-gnu/cmake/pcl`. So, It can't find correct PCL include directory(the correct PCL include dir is `/usr/include/pcl-1.10`) and lead to cmake error.**

**My Environment :**

 - OS: [Ubuntu 20.04]
 - Compiler: [GCC 9.3.0]
 - PCL Version [1.10 ]
 - PCL Type: [Installed]
 **some more info about pcl**
```
Package: libpcl-dev
Version: 1.10.0+dfsg-5ubuntu1
Priority: extra
Section: universe/libdevel
Source: pcl
Origin: Ubuntu
```
**So, what's the problem with my PCL, and how can I find PCL by `find_package(PCL REQUIRED)` successfully? Thanks a lot!**

### PR DESCRIPTION
**Pull request by perisage**
Fixes #4661 
Make PCLConfig.cmake go well on some special platform described as #4661.

### CODE REVIEW

**Review by kunaltyagi — DISMISSED** [review]
Waiting for tutorial CI. If that's red, please reject this approval

**Comment by perisage** [comment]
> Waiting for tutorial CI. If that's red, please reject this approval


Unfortunately, the tutorial CI failed. After checking the log while building tutorial CI,  I think I found the reason.

There is a difference between the `PCLConfig.cmake` which corresponds to the PCL installed by `sudo apt install libpcl-dev` and the `PCLConfig.cmake.in` in the source code.
The code for obtaining `PCL_ROOT` in the `PCLConfig.cmake` file of the former situation(install via `apt`) is as follows(from my system under `/usr/lib/x86_64-linux-gnu/cmake/pcl/PCLConfig.cmake`):
```cmake
if(WIN32 AND NOT MINGW)
# PCLConfig.cmake is installed to PCL_ROOT/cmake
  get_filename_component(PCL_ROOT "${PCL_DIR}" PATH)
else()
# PCLConfig.cmake is installed to PCL_ROOT/share/pcl-x.y
  get_filename_component(PCL_ROOT "${CMAKE_CURRENT_LIST_DIR}/../../../.." ABSOLUTE)
endif()
```
And the code for obtaining `PCL_ROOT` in the source code `PCLConfig.cmake.in` file is as follows:
```cmake
 if(WIN32 AND NOT MINGW) 
 # PCLConfig.cmake is installed to PCL_ROOT/cmake 
   get_filename_component(PCL_ROOT "${PCL_DIR}" PATH) 
 else() 
 # PCLConfig.cmake is installed to PCL_ROOT/share/pcl-x.y 
   get_filename_component(PCL_ROOT "${CMAKE_CURRENT_LIST_DIR}/../.." ABSOLUTE) 
 endif() 
```
Through the above code, we can find that the former returns to the upper-level directory **4 times** on the basis of `CMAKE_CURRENT_LIST_DIR`. everything is ok on my system after getting `REALPATH` because the installation location of `PCLConfig.cmake`  at this time is `/usr/lib/x86_64-linux-gnu/cmake/pcl/PCLConfig.cmake`.


But the latter(source code in `PCLConfig.cmake.in` ) only returns to the upper-level directory **twice**. 


Based on this, I checked the log of the tutorial CI in the process of building and found that PCL is installed from source code. The source code directory is `/home/vsts/work/1/s/`(the corresponding log is [This](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=18433&view=logs&j=e39aa7be-e494-54ce-b7c7-9d3a57135baa&t=eb16b0a1-a5e8-5682-7c18-90824d4907db&l=14)).


The build directory is `/__w/1/build/`((the corresponding log is [This](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=18433&view=logs&j=e39aa7be-e494-54ce-b7c7-9d3a57135baa&t=cf06a82e-f9e6-5a04-b4dc-e71667809f39&l=146)).


 The install directory is `/__w/1/install/` (the corresponding log is [This](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=18433&view=logs&j=e39aa7be-e494-54ce-b7c7-9d3a57135baa&t=db3f18d8-2a02-5246-567b-32add05a15d1&l=33)). The PCLConfig file was installed in `/__w/1/install/share/pcl-1.11/PCLConfig.cmake`  (the corresponding log is [This](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=18433&view=logs&j=e39aa7be-e494-54ce-b7c7-9d3a57135baa&t=db3f18d8-2a02-5246-567b-32add05a15d1&l=50))


Therefore, system can find the PCL `include` folder(`/__w/1/install/include`) after returning to the upper-level directory **twice** not **four times**.

So, should I change the 
```cmake
get_filename_component(PCL_ROOT "${CMAKE_CURRENT_LIST_DIR}/../../../.." ABSOLUTE)
```
 in `PCLConfig.cmake.in` to 
```cmake
  get_filename_component(PCL_CONFIG_PATH "${CMAKE_CURRENT_LIST_DIR}" REALPATH)
  get_filename_component(PCL_ROOT "${PCL_CONFIG_PATH}/../.." ABSOLUTE)
```
and let it pass CI.
But what about the case of installing in `/usr/lib/x86_64-linux-gnu/cmake/pcl/PCLConfig.cmake` via `apt`?

**Comment by kunaltyagi** [comment]
In that case, please go ahead with your recommended changes, and raise a big report with the package maintainers. They'll add a patch and re-release it sometime :)

This update will be released in some newer release of debian and ubuntu

**Comment by mvieth** [comment]
If you look here: https://launchpad.net/ubuntu/+archive/primary/+sourcefiles/pcl/1.10.0+dfsg-5ubuntu1/pcl_1.10.0+dfsg-5ubuntu1.debian.tar.xz
under debian/patches/0001-Fix... there is a patch that does this:
```diff
-  get_filename_component(PCL_ROOT "${CMAKE_CURRENT_LIST_DIR}/../.." ABSOLUTE)
+  get_filename_component(PCL_ROOT "${CMAKE_CURRENT_LIST_DIR}/../../../.." ABSOLUTE)
```

**Pull request closed without merging**
